feat(pie-list): DSW-2960 implement list + list item component and tests

matthewhardern opened the pull request:
## Describe your changes (can list changeset entries if preferable)

Implements the initial version of the `pie-list` and `pie-list-item` components with basic functionality:

**pie-list component:**
- Added `variant` prop supporting 'default' and 'compact' layouts
- Added `dividers` prop to show separators between list items
- Implemented proper semantic HTML with `<ul role="list">`
- Integrated with PIE design system using CSS custom properties

**pie-list-item component:**
- Added `primaryText` prop for text content
- Added `selected` and `disabled` state props
- Implemented proper accessibility with ARIA attributes
- Applied PIE typography using font helpers

**Additional changes:**
- Created Storybook stories with 4 variants (Default, Compact, WithDividers, CompactWithDividers)
- Added test helpers and page objects for component testing
- Implemented RTL support via RtlMixin
- Added proper data-test-id attributes for testing

## Author Checklist (complete before requesting a review, do not delete any)
- [x] I have performed a self-review of my code.
- [x] I have added thorough tests where applicable (unit / component / visual).
- [x] I have reviewed the `PIE Storybook`/`PIE Docs` PR preview.
- [x] I have reviewed visual test updates properly before approving.
- [x] If changes will affect consumers of the package, I have created a changeset entry.
- [x] If a changeset file has been created, I have tested these changes in [PIE Aperture](https://github.com/justeattakeaway/pie-aperture/) using the `/test-aperture` command.
- [x] I have filled out the DS Review Tracker checklist (Moving PR to "Ready to Review")

## Not-applicable Checklist items
- Visual test updates (none in this PR)

| Task                   | Link                             |
|------------------------|----------------------------------|
| Aperture PR            | [🔗](#) |
| NextJS 14 deployment   | [🔗](#) |
| Nuxt 3 deployment      | [🔗](#) |
| Vanilla deployment     | [🔗](#) |

## Reviewer checklists (complete before approving)
Mark items as `[-] N/A` if not applicable.

### Reviewer 1 - @jamieomaguire 
- [x] I have reviewed the `PIE Storybook`/`PIE Docs` PR preview.
- [x] I have verified that all acceptance criteria for this ticket have been completed.
- [x] I have reviewed the Aperture changes (if added)
- [x] If there are visual test updates, I have reviewed them.

### Reviewer 2 - @ashleynolan 
- [x] I have reviewed the `PIE Storybook`/`PIE Docs` PR preview.
- [x] I have verified that all acceptance criteria for this ticket have been completed.
- [x] I have reviewed the Aperture changes (if added)
- [x] If there are visual test updates, I have reviewed them.

Commit:
feat(pie-list): implement list item component and tests

Adds styles, page objects, and selectors for the pie-list-item
component. Implements interactive states, layout, and
accessibility features. Introduces test helpers for easier
component interaction and attribute checking.

 - pie-list component:
    - variant prop (default, compact)
    - dividers prop
    - Proper semantic HTML (<ul role="list">)
    - Design system integration with CSS custom properties
  - pie-list-item component:
    - primaryText prop
    - selected and disabled states
    - Proper accessibility (ARIA attributes)
    - Correct typography using pie-css font helpers
  - Storybook integration:
    - 4 working story variants
    - Interactive controls
    - Visual verification available

Commit:
refactor(pie-list): Update list item rendering and behavior

Remove 'role' attribute and adjust primary text rendering logic.
Update data-test-id for list item selector. removing some stuff from this PR

Commit:
feat: Update pie-list Storybook stories and controls

Add new stories for compact and divider variants of pie-list
component. Implement argTypes for variant and dividers props.
Update Template to use variant and dividers props. Include
pie-list-item import and add sample items to the list.

Commit:
test: Update PieList component tests and remove unused code

Updates PieList component tests to use correct story ID and adds
new tests for variant and dividers. Removes unused code in
pie-list-item and ListPage. Improves test coverage and removes
unnecessary attributes.

matthewhardern commented on packages/components/pie-list/src/list.scss:
```diff
@@ -1,5 +1,79 @@
+  direction: rtl;
+}
+
+.c-list {
+  display: block;
+  margin: 0;
+  padding: 0;
+  list-style: none;
+  width: 100%;
+
+  // Default padding - will be inherited by child list items
+  --list-item-padding: var(--dt-spacing-d) var(--dt-spacing-d) var(--dt-spacing-d) var(--dt-spacing-d);
+
+  // Remove default browser list padding/margin
+  &:where(ul, ol) {
+    margin: 0;
+    padding: 0;
+  }
+}
+
+// Variant: Compact
+.c-list--compact {
+  --list-item-padding: var(--dt-spacing-b) var(--dt-spacing-d) var(--dt-spacing-b) var(--dt-spacing-d);
+}
+
+// Interactive items
+.c-list--interactive ::slotted(*) {
+  cursor: pointer;
+}
+
+// Dividers between items
+.c-list--dividers ::slotted(*:not(:last-child)) {
+  --list-item-border-bottom: 1px solid var(--dt-color-border-subtle);
+
+  border-bottom: 1px solid var(--dt-color-border-subtle);
+}
```
--list-item-padding: var(**--dt-spacing-b**) var(--dt-spacing-d) var(**--dt-spacing-b**) var(--dt-spacing-d);
is not equal to
--list-item-padding: var(--dt-spacing-d) var(--dt-spacing-d) var(--dt-spacing-d) var(--dt-spacing-d);

matthewhardern commented on packages/components/pie-list/src/pie-list-item/index.ts:
```diff
@@ -1,15 +1,64 @@
+
+    render () {
+        const classes = {
+            'c-list-item': true,
+            'c-list-item--selected': this.selected,
+            'c-list-item--disabled': this.disabled,
+            'c-list-item--rtl': this.isRTL,
```
thanks will make the required changes

siggerzz commented on packages/components/pie-list/test/component/pie-list.spec.ts:
```diff
@@ -1,20 +1,51 @@
+    });
+
+    test('should render with variant="compact"', async ({ page }) => {
+        // Arrange
+        const listPage = new BasePage(page, 'list--default');
+        const props: ListProps = {
+            variant: 'compact',
+        };
+
+        await listPage.load({ ...props });
+
+        // Act
+        const listComponent = page.locator('pie-list');
+
+        // Assert
+        await expect(listComponent).toBeVisible();
+        await expect(listComponent).toHaveAttribute('variant', 'compact');
+    });
+
+    test('should render with dividers enabled', async ({ page }) => {
+        // Arrange
+        const listPage = new BasePage(page, 'list--default');
+        const props: ListProps = {
+            dividers: true,
+        };
 
-        basePage.load();
-        await page.waitForTimeout(2500);
+        await listPage.load({ ...props });
 
         // Act
-        const list = page.locator(componentSelector);
+        const listComponent = page.locator('pie-list');
 
         // Assert
-        expect(list).toBeVisible();
+        await expect(listComponent).toBeVisible();
+        await expect(listComponent).toHaveAttribute('dividers');
     });
 });
```
Currently you have the following test stories:

```
export const Default = createStory<ListProps>(Template, defaultArgs)();

export const Compact = createStory<ListProps>(Template, {
    ...defaultArgs,
    variant: 'compact',
})();

export const WithDividers = createStory<ListProps>(Template, {
    ...defaultArgs,
    dividers: true,
})();

export const CompactWithDividers = createStory<ListProps>(Template, {
    ...defaultArgs,
    variant: 'compact',
    dividers: true,
})();
```

Given you're just using the default story in the test spec, you can either:
- Remove the unused stories
- Refactor the tests to navigate to the other stories, meaning you no longer have to provide the `props` object in the test (as it'll be set by default in the Story).

siggerzz commented:
Is it possible to add some a11y / visual tests for this component?

For visual tests, we leverage the `createVariantsStory` that allows you to render all prop variations on a single story. you can see an example here:

Story - https://github.com/justeattakeaway/pie/blob/main/apps/pie-storybook/stories/testing/pie-assistive-text.test.stories.ts#L66-L72
Test Spec - https://github.com/justeattakeaway/pie/blob/main/packages/components/pie-assistive-text/test/visual/pie-assistive-text.spec.ts#L9-L17 (This also includes logic for RTL)

Commit:
refactor(pie-list): Update class names and remove RTL support

Updates class names for consistency and removes RTL-specific
classes. Simplifies list styles and updates changeset message.
Improves overall component structure and naming conventions.

Commit:
test: Update PieList component tests and stories

Refactor tests to use specific story variants
Add new test for compact variant with dividers
Update stories to include variants and dividers
Remove unused import and simplify test setup

Commit:
feat(pie-list): DSW-2963 host-based semantics and hasDividers prop

- expose hasDividers boolean on the list host and default props
- drop internal <ul>/<li> wrappers so role/aria live on the custom elements
- update SCSS, stories, and Playwright specs to reflect host attributes

Commit:
feat(pie-list): DSW-2960 support ordered lists

# Conflicts:
#	packages/components/pie-list/pie-list.test.stories.ts
#	packages/components/pie-list/test/component/pie-list.spec.ts

Commit:
feat(pie-list): DSW-2960 support ordered lists

matthewhardern committed:
Merge branch 'main' into dsw-2960-basic-list

jamieomaguire commented on packages/components/pie-list/src/index.ts:
```diff
@@ -12,11 +18,50 @@ const componentSelector = 'pie-list';
 /**
  * @tagname pie-list
+ *
+ * PIE List component for displaying collections of related items.
```
we will want to document the slot(s) in this comment too - you can see examples in other components for the correct formatting

jamieomaguire commented on packages/components/pie-list/src/list.scss:
```diff
@@ -1,5 +1,37 @@
 :host {
-    display: block;
+  display: block;
+  width: 100%;
```
is width 100% needed for a block element?

jamieomaguire commented on packages/components/pie-list/src/list.scss:
```diff
@@ -1,5 +1,37 @@
+
+// Dividers between items
+:host([has-dividers]) ::slotted(pie-list-item:not(:last-child)),
+:host([has-dividers]) ::slotted(li:not(:last-child)) {
```
are we planning on using `li` elements in the slot?

jamieomaguire commented on packages/components/pie-list/src/pie-list-item/index.ts:
```diff
@@ -1,15 +1,76 @@
+        }
+
+        if (!this.hasAttribute('data-test-id')) {
+            this.setAttribute('data-test-id', 'pie-list-item');
```
I understand that this only happens if none is provided by consumers, but it would lead to duplicate test-ids. Unsure if we should add a `uuid` or something (or not self-render a test ID at all?

matthewhardern committed:
Merge branch 'main' into dsw-2960-basic-list

ashleynolan committed:
Merge branch 'main' into dsw-2960-basic-list

siggerzz commented on packages/components/pie-list/test/helpers/page-object/pie-list.component.ts:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Checks if the list has the specified attribute
+     */
+    async hasAttribute (page: Page, attribute: string): Promise<boolean> {
+        return page.evaluate(
+            ([selector, attr]) => {
+                const element = document.querySelector(selector);
+                return element?.hasAttribute(attr) ?? false;
+            },
+            ['pie-list', attribute],
+        );
+    }
```
Playwright has a native method for this.

Something in the test spec like `await expect(foo).toHaveAttribute('bar');` should do the trick 👍🏼

matthewhardern committed:
Merge branch 'main' into dsw-2960-basic-list

Commit:
test(pie-list): DSW-2960 remove redundant attribute helper

raoufswe commented on apps/pie-storybook/stories/pie-list.stories.ts:
```diff
@@ -1,19 +1,36 @@
     component: 'pie-list',
-    argTypes: {},
+    argTypes: {
+        variant: {
+            control: 'select',
+            options: ['default', 'compact'],
+        },
+        hasDividers: {
+            control: 'boolean',
+        },
+        listType: {
+            control: 'select',
+            options: ['unordered', 'ordered'],
+        },
```
Let's use the exported defaults here instead of hardcoding them as hardcoding tends to get out of sync easily.

[example](https://github.com/justeattakeaway/pie/blob/main/apps/pie-storybook/stories/pie-card.stories.ts#L48-L51)
the changes would look like: 

```suggestion
        variant: {
            control: 'select',
            options: variants,
                defaultValue: {
                summary: defaultProps.variant,
            },
        }
```

raoufswe commented on apps/pie-storybook/stories/testing/pie-list.test.stories.ts:
```diff
@@ -1,19 +1,36 @@
     title: 'List',
     component: 'pie-list',
-    argTypes: {},
+    argTypes: {
+        variant: {
+            control: 'select',
+            options: ['default', 'compact'],
+        },
+        hasDividers: {
+            control: 'boolean',
+        },
+        listType: {
+            control: 'select',
+            options: ['unordered', 'ordered'],
+        },
+    },
```
similar suggestion [here](https://github.com/justeattakeaway/pie/pull/2404/files#r2405556390)

raoufswe commented on packages/components/pie-list/pie-list.test.stories.ts:
```diff
@@ -25,10 +25,36 @@ const listStoryMeta: ListStoryMeta = {
+const Template = ({ variant, hasDividers, listType }: ListProps) => html`
+    <pie-list
+        variant=${variant}
+        list-type=${listType ?? 'unordered'}
+        ?has-dividers=${hasDividers}>
```
```suggestion
        listType=${listType ?? 'unordered'}
        ?hasDividers=${hasDividers}>
```

Commit:
docs(pie-list): DSW-2960 sync stories with exported defaults

raoufswe commented on packages/components/pie-list/src/index.ts:
```diff
@@ -12,11 +18,53 @@ const componentSelector = 'pie-list';
+    connectedCallback () {
+        super.connectedCallback();
+
+        if (!this.hasAttribute('data-test-id')) {
+            this.setAttribute('data-test-id', 'pie-list');
+        }
```
could you please share more why we need to do this step? i think it's common for users to set a custom data-test-id on the host element and that will always override it

raoufswe commented on packages/components/pie-list/src/index.ts:
```diff
@@ -12,11 +18,53 @@ const componentSelector = 'pie-list';
+    @property({ type: String, reflect: true })
+    public variant: ListVariant = defaultProps.variant;
+
+    @property({
+        type: Boolean,
+        reflect: true,
+        attribute: 'has-dividers',
+    })
+    public hasDividers = defaultProps.hasDividers;
+
+    @property({
+        type: String,
+        reflect: true,
+        attribute: 'list-type',
+    })
+    public listType: ListType = defaultProps.listType;
```
yeah makes sense for the styles part! Sorry I haven’t had the chance to review the file yet.
About the conversion, I think you can still use the prop name in camelCase as used [here](https://github.com/justeattakeaway/pie/blob/main/packages/components/pie-breadcrumb/src/breadcrumb.scss#L75). I'm quite reluctant about the transformation since all props in the codebase currently use the default Lit transformation, and mixing approaches might be confusing for consumers. it would also be a breaking change if we decided to standardize this approach for existing components so we need to be careful. As far as I know, all frameworks and CSS still understand camelCase props, so i would give it a try to avoid doing so unless necessary.

raoufswe commented on packages/components/pie-list/src/pie-list-item/index.ts:
```diff
@@ -1,15 +1,81 @@
+            <div class="c-listItem-content">
+                <div class="c-listItem-primary">
+                    <slot></slot>
+                    ${!this.querySelector(':not([slot])') && this.primaryText ? html`<span>${this.primaryText}</span>` : ''}
```
this will probably break in SSR and result in a hydration mismatch. have we tested in aperture?

Commit:
fix(pie-list): DSW-2960 use host classes for state styling

Commit:
fix(pie-list): DSW-2960 use host classes for state styling

matthewhardern commented on packages/components/pie-list/src/index.ts:
```diff
@@ -12,11 +18,53 @@ const componentSelector = 'pie-list';
 @safeCustomElement('pie-list')
 export class PieList extends RtlMixin(PieElement) implements ListProps {
+    @property({ type: String, reflect: true })
+    public variant: ListVariant = defaultProps.variant;
```
nice to know, I have now followed this practice

raoufswe commented on packages/components/pie-list/src/pie-list-item/list-item.scss:
```diff
@@ -1 +1,67 @@
+  // These variables can be overridden by the parent list component
+  --list-item-leading-margin: var(--dt-spacing-c);
+  --list-item-trailing-margin: var(--dt-spacing-c);
+  --list-item-hover-bg: var(--dt-color-hover-1);
+  --list-item-active-bg: var(--dt-color-active-1);
```
doesn't seem used yet but the tokens should be 

```suggestion
  --list-item-hover-bg: var(--dt-color-hover-01);
  --list-item-active-bg: var(--dt-color-active-01);
```

Commit:
Merge remote-tracking branch 'origin/dsw-2960-basic-list' into dsw-2960-basic-list

Commit:
fix(pie-list): DSW-2960 rely on slot fallback and drop default id

Commit:
fix(pie-list): DSW-2960 use active-01 tokens

raoufswe commented on packages/components/pie-list/pie-list.test.stories.ts:
```diff
@@ -2,13 +2,14 @@ import { html } from 'lit';
 import { type Meta } from '@storybook/web-components';
```
this file is not needed and should be removed as we don't add stories in component folders.

raoufswe commented on apps/pie-storybook/stories/testing/pie-list.test.stories.ts:
```diff
@@ -25,10 +52,44 @@ const listStoryMeta: ListStoryMeta = {
+    <pie-list
+        .variant=${variant}
+        .listType=${listType}
+        .hasDividers=${hasDividers}>
```
the dot syntax is needed for non primitive values such as arrays or objects or if we need to set it via js. is there a reason why we are using that? otherwise i suggest to use


```suggestion
        variant=${ifDefined(variant)}
        listType=${ifDefined(listType)}
        ?hasDividers=${hasDividers}>
```

Commit:
fix(pie-list): DSW-2960 migrate to classMap styling and update templates

Commit:
fix(pie-list): DSW-2960 align pie-list attribute handling and tests

matthewhardern committed:
Merge branch 'main' into dsw-2960-basic-list

Commit:
chore(pie-list): DSW-2960 remove component test suite

matthewhardern committed:
Merge branch 'main' into dsw-2960-basic-list

ashleynolan committed:
Merge branch 'main' into dsw-2960-basic-list

raoufswe commented on packages/components/pie-list/src/pie-list-item/index.ts:
```diff
@@ -1,15 +1,91 @@
+
+    private _syncHostClasses () {
+        this.classList.toggle('is-selected', this.isSelected);
+    }
```
is there a reason why we can't reflect the `isSelected` prop and use it for styles such as 

```
 @property({ type: Boolean, reflect: true})
    public isSelected = defaultProps.isSelected;

```
and style it as 

```
:host(isSelected) .c-listItem-primary span {
  color: var(--dt-color-content-default);
}
```

Commit:
fix(pie-list): DSW-2960 reflect isSelected and Lit bindings

matthewhardern committed:
Update packages/components/pie-list/src/pie-list-item/list-item.scss

Co-authored-by: Raouf <raouf.sawehli@justeattakeaway.com>

matthewhardern committed:
Update packages/components/pie-list/src/pie-list-item/list-item.scss

Co-authored-by: Raouf <raouf.sawehli@justeattakeaway.com>

matthewhardern committed:
Merge branch 'main' into dsw-2960-basic-list

raoufswe commented on packages/components/pie-list/src/list.scss:
```diff
@@ -1,5 +1,35 @@
+}
+
+// Dividers between items
+:host([has-dividers]) ::slotted(pie-list-item:not(:last-child)),
```
```suggestion
:host([hasDividers]) ::slotted(pie-list-item:not(:last-child)),
```

raoufswe commented on packages/components/pie-list/src/list.scss:
```diff
@@ -1,5 +1,35 @@
+// Dividers between items
+:host([has-dividers]) ::slotted(pie-list-item:not(:last-child)),
+// Allow bespoke markup that renders native `<li>` elements.
+:host([has-dividers]) ::slotted(li:not(:last-child)) {
```
```suggestion
:host([hasDividers]) ::slotted(li:not(:last-child)) {
```
